hummingbot/strategy/perpetual_market_making/perpetual_market_making.py

```diff
@@ -0,0 +1,975 @@
+                price = market.quantize_order_price(self.trading_pair, take_profit_price)
+                size = market.quantize_order_amount(self.trading_pair, abs(position.amount))
+                old_exit_orders = [
+                    o for o in self.active_orders
+                    if ((o.price != price or o.quantity != size)
+                        and o.client_order_id in self._exit_orders.keys()
+                        and ((position.amount < 0 and o.is_buy) or (position.amount > 0 and not o.is_buy)))]
+                for old_order in old_exit_orders:
+                    self.cancel_order(self._market_info, old_order.client_order_id)
+                    self.logger().info(
+                        f"Initiated cancellation of previous take profit order {old_order.client_order_id} in favour of new take profit order.")
+                exit_order_exists = [o for o in self.active_orders if o.price == price]
+                if len(exit_order_exists) == 0:
+                    if size > 0 and price > 0:
+                        if position.amount < 0:
+                            buys.append(PriceSize(price, size))
+                        else:
+                            sells.append(PriceSize(price, size))
+        return Proposal(buys, sells)
+
+    def _should_renew_stop_loss(self, stop_loss_order: LimitOrder) -> bool:
+        stop_loss_creation_timestamp = self._exit_orders.get(stop_loss_order.client_order_id)
+        time_since_stop_loss = self.current_timestamp - stop_loss_creation_timestamp
+        return time_since_stop_loss >= self._time_between_stop_loss_orders
+
+    def stop_loss_proposal(self, mode: PositionMode, active_positions: List[Position]) -> Proposal:
+        market: ExchangeBase = self._market_info.market
+        top_ask = market.get_price(self.trading_pair, False)
+        top_bid = market.get_price(self.trading_pair, True)
+        buys = []
+        sells = []
+
+        for position in active_positions:
+            # check if stop loss order needs to be placed
+            stop_loss_price = position.entry_price * (Decimal("1") + self._stop_loss_spread) if position.amount < 0 \
+                else position.entry_price * (Decimal("1") - self._stop_loss_spread)
+            existent_stop_loss_orders = [order for order in self.active_orders
+                                         if order.client_order_id in self._exit_orders.keys()
+                                         and ((position.amount > 0 and not order.is_buy)
+                                              or (position.amount < 0 and order.is_buy))]
+            if (not existent_stop_loss_orders
+                    or (self._should_renew_stop_loss(existent_stop_loss_orders[0]))):
+                previous_stop_loss_price = None
+                for order in existent_stop_loss_orders:
+                    previous_stop_loss_price = order.price
+                    self.cancel_order(self._market_info, order.client_order_id)
```

Comment:
Is there a possibility of multiple stop loss orders for a position?

Reply:
I discussed this scenario with @phbrgnomo. If a previous stop loss is still alive, it means that the current price has moved away from it and it can no longer fulfill its intention. That is why we are cancelling it before creating the new one.